File: backend/app/schemas/priority_queue.py
"""Read-only priority outreach queue API schemas."""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class PriorityStudentBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    gmail_address: str
    status: str = "active"
    email_health_status: str = "healthy"
    is_demo: bool = False


class PriorityHRBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    company: str
    email: str
    is_valid: bool = True
    status: str | None = None


class PriorityQueueRow(BaseModel):
    student: PriorityStudentBrief
    hr: PriorityHRBrief
    priority_score: float = Field(..., ge=0, le=100)
    priority_rank: int = Field(..., ge=1)
    recommendation_reason: list[str] = Field(default_factory=list)
    recommended_action: str
    urgency_level: str
    queue_bucket: str
    hr_tier: str
    health_score: float
    opportunity_score: float
    dimension_scores: dict[str, float] = Field(default_factory=dict)
    next_best_touch: str | None = None
    cooldown_status: str | None = None
    followup_status: str | None = None
    campaign_id: UUID | None = None
    signal_fingerprint: str
    ranking_mode: str = "standard"
    ranking_slot_type: str | None = None
    diversity_note: str | None = None
    decision_diagnostics: dict[str, Any] = Field(
        default_factory=dict,
        description="Operator-facing explainability: rank drivers, suppression, follow-up snapshot, cooldown, scoring.",
    )


class PriorityQueueSummary(BaseModel):
    send_now_count: int = 0
    followup_due_count: int = 0
    warm_lead_priority_count: int = 0
    wait_for_cooldown_count: int = 0
    suppressed_count: int = 0
    low_priority_count: int = 0
    avg_priority_score: float | None = None
    total_candidates: int = 0


class PriorityQueueResponse(BaseModel):
    computed_at_utc: datetime
    summary: PriorityQueueSummary
    rows: list[PriorityQueueRow]
    diversity_metrics: dict[str, Any] = Field(default_factory=dict)


class PriorityQueueSummaryOnlyResponse(BaseModel):
    computed_at_utc: datetime
    summary: PriorityQueueSummary


class SchedulerPriorityHookDoc(BaseModel):
    """Design-only: when enabled, scheduler may prefer higher-ranked pairs (Phase 2+)."""

    env_flag: str = "SCHEDULER_USE_PRIORITY_QUEUE"
    enabled_in_environment: bool = False
    description: str = (
        "If set to true, future scheduler iterations may reorder due campaigns using "
        "the read-only priority engine. Phase 1 ships with this disabled; no sending behavior changes."
    )

File: frontend/src/api/priorityQueue.ts
import { safeGet } from "@/api/api";

export type QueueBucket =
  | "SEND_NOW"
  | "FOLLOW_UP_DUE"
  | "WARM_LEAD_PRIORITY"
  | "WAIT_FOR_COOLDOWN"
  | "SUPPRESS"
  | "LOW_PRIORITY";

export type PriorityQueueSummary = {
  send_now_count: number;
  followup_due_count: number;
  warm_lead_priority_count: number;
  wait_for_cooldown_count: number;
  suppressed_count: number;
  low_priority_count: number;
  avg_priority_score: number | null;
  total_candidates: number;
};

export type PriorityStudentBrief = {
  id: string;
  name: string;
  gmail_address: string;
  status?: string;
  email_health_status?: string;
  is_demo?: boolean;
};

export type PriorityHRBrief = {
  id: string;
  name: string;
  company: string;
  email: string;
  is_valid?: boolean;
  status?: string | null;
};

export type FollowUpDiagnosticSnapshot = {
  status: string | null;
  eligible_for_followup: boolean;
  blocked_reason: string | null;
  next_followup_step: number | null;
  next_template_type: string | null;
  due_date_utc: string | null;
  days_until_due: number | null;
  paused: boolean;
  send_in_progress: boolean;
  initial_or_anchor_campaign_id: string | null;
};

/** Nested explainability payload from GET /queue/priority (read-only diagnostics). */
export type DecisionDiagnostics = {
  decision_computed_at_utc: string;
  last_pair_activity_utc: string | null;
  queue_bucket: string;
  bucket_rationale: string;
  recommended_action: string;
  why_ranked: string[];
  why_suppressed: string[];
  follow_up: FollowUpDiagnosticSnapshot;
  cooldown: {
    summary_line: string | null;
    penalty_reasons: string[];
    cooldown_penalty_score: number;
  };
  scoring: {
    priority_score: number;
    blended_before_cooldown_subtraction: number;
    cooldown_subtracted: number;
    formula: string;
    axes: { name: string; value: number; weight: number; weighted: number }[];
    top_components: { name: string; value: number; weight: number; weighted: number }[];
  };
  why_not_sent: {
    is_suppressed: boolean;
    summary: string;
    blockers: string[];
    all_signal_lines: string[];
    follow_up_snapshot: FollowUpDiagnosticSnapshot;
    operator_note: string;
  } | null;
  waiting_or_deferred: {
    bucket_is_wait: boolean;
    negative_signals: string[];
    summary: string;
  } | null;
};

export type PriorityQueueRow = {
  student: PriorityStudentBrief;
  hr: PriorityHRBrief;
  priority_score: number;
  priority_rank: number;
  recommendation_reason: string[];
  recommended_action: string;
  urgency_level: string;
  queue_bucket: QueueBucket;
  hr_tier: string;
  health_score: number;
  opportunity_score: number;
  dimension_scores: Record<string, number>;
  next_best_touch: string | null;
  cooldown_status: string | null;
  followup_status: string | null;
  campaign_id: string | null;
  signal_fingerprint: string;
  ranking_mode?: string;
  ranking_slot_type?: string | null;
  diversity_note?: string | null;
  decision_diagnostics?: DecisionDiagnostics;
};

export type PriorityQueueResponse = {
  computed_at_utc: string;
  summary: PriorityQueueSummary;
  rows: PriorityQueueRow[];
  diversity_metrics?: Record<string, unknown>;
};

const STORAGE_KEY = "priorityQueuePrev.v1";

export type StoredQueueEntry = { rank: number; fingerprint: string };

export function loadPrevQueueMap(): Record<string, StoredQueueEntry> {
  try {
    const raw = sessionStorage.getItem(STORAGE_KEY);
    if (!raw) return {};
    const p = JSON.parse(raw) as unknown;
    return p && typeof p === "object" ? (p as Record<string, StoredQueueEntry>) : {};
  } catch {
    return {};
  }
}

export function saveQueueSnapshot(rows: PriorityQueueRow[]) {
  const m: Record<string, StoredQueueEntry> = {};
  for (const r of rows) {
    const k = `${r.student.id}:${r.hr.id}`;
    m[k] = { rank: r.priority_rank, fingerprint: r.signal_fingerprint };
  }
  sessionStorage.setItem(STORAGE_KEY, JSON.stringify(m));
}

export function rowPairKey(r: PriorityQueueRow) {
  return `${r.student.id}:${r.hr.id}`;
}

export function fetchPriorityQueue(params?: {
  bucket?: string;
  student_id?: string;
  tier?: string;
  only_due?: boolean;
  limit?: number;
  include_demo?: boolean;
  diversified?: boolean;
}): Promise<PriorityQueueResponse> {
  return safeGet<PriorityQueueResponse>("/queue/priority", params as Record<string, unknown> | undefined);
}

export function fetchPriorityQueueSummary(params?: {
  bucket?: string;
  student_id?: string;
  tier?: string;
  only_due?: boolean;
  limit?: number;
  include_demo?: boolean;
}): Promise<{ computed_at_utc: string; summary: PriorityQueueSummary }> {
  return safeGet("/queue/priority/summary", params as Record<string, unknown> | undefined);
}
